### PR TITLE
[tcp-close] New whocloses semantics

### LINKED ISSUE #526: [test] Integration Tests for TCP `close()`
## Context

We should have the following integration tests for TCP `close()`:

- [x] Attempt to close an invalid socket. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket multiple times. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket that is not bound. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket that is bound. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket that is listening. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket that is accepting. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [x] Attempt to close a TCP socket that is connecting. 0cc11607cba5b7eef0ee4778e3839ca698cdaa7c
- [ ] Attempt to close a TCP socket that is connected.

### PR DESCRIPTION
**Pull request by anandbonde**
Closes #526

### CODE REVIEW

File: examples/tcp-close/client.rs [hunk 1]
```diff
@@ -150,4 +151,124 @@ impl TcpClient {
...
+            // Connect to the server and wait.
+            let sockqd: QDesc = self.libos.socket(AF_INET, SOCK_STREAM, 0)?;
+            let qt: QToken = self.libos.connect(sockqd, self.remote)?;
+            let qr: demi_qresult_t = self.libos.wait(qt, None)?;
+
+            match qr.qr_opcode {
+                demi_opcode_t::DEMI_OPC_CONNECT => {
+                    println!("{} clients connected", i + 1);
+
+                    // pop immediately after connect and wait
```
Comment (ppenna) [inline examples/tcp-close/client.rs:168]: Could you add a full step after the comment?

File: examples/tcp-close/client.rs [hunk 2]
```diff
@@ -150,4 +151,124 @@ impl TcpClient {
...
+            // Close TCP socket.
+            self.libos.close(sockqd)?;
+        }
+
+        Ok(())
+    }
+
+    /// Attempts to make several connections concurrently.
+    pub fn run_concurrent_expecting_server_to_close_sockets(&mut self, num_clients: usize) -> Result<()> {
+        let mut q_descs: Vec<QDesc> = Vec::default();
```
Comment (ppenna) [inline examples/tcp-close/client.rs:209]: Our code base uses `qds` for naming this variable. Could you change your code to match the existing pattern?

File: examples/tcp-close/client.rs [hunk 3]
```diff
@@ -150,4 +151,124 @@ impl TcpClient {
...
+        let mut q_tokens: Vec<QToken> = Vec::default();
```
Comment (ppenna) [inline examples/tcp-close/client.rs:210]: Our code base uses `qts` for naming this variable. Could you change your code to match the existing pattern?

File: examples/tcp-close/client.rs [hunk 4]
```diff
@@ -150,4 +151,124 @@ impl TcpClient {
...
+
+            let qr: demi_qresult_t = {
+                let (index, qr): (usize, demi_qresult_t) = self.libos.wait_any(&q_tokens, None)?;
+                let _qt: QToken = q_tokens.remove(index);
+                qr
+            };
+
+            match qr.qr_opcode {
+                demi_opcode_t::DEMI_OPC_CONNECT => {
+                    let sock_qd: QDesc = qr.qr_qd.into();
```
Comment (ppenna) [inline examples/tcp-close/client.rs:235]: Our code base use `sockqd` for naming this variable. Could you change your code to match the existing pattern?

**Comment by ppenna** [comment]
@anandbonde could you change the description of this issue so that it links issues correctly? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Review by ppenna — COMMENTED** [review]
Tests are not passing. Why? Could you file an issue to keep track of this bug? If you disable this test for some LibOSes, could you file an issue to keep track which LibOSes fall in this condition?

**Comment by anandbonde** [comment]
> @anandbonde could you change the description of this issue so that it links issues correctly? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue



> @anandbonde could you change the description of this issue so that it links issues correctly? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Done

**Comment by anandbonde** [comment]
> Tests are not passing. Why? Could you file an issue to keep track of this bug? If you disable this test for some LibOSes, could you file an issue to keep track which LibOSes fall in this condition?

Just checked the logs, the tests are failing because I have added a new argument to the tcp-close utility, but the ci scripts are not updated to use it. I think it would be good to change the ci scripts to be updated in this PR so that we don't have to revisit this area. Do you have any preference?

**Comment by ppenna** [comment]
Please, update the CI scripts accordingly. I am also confident that this will not work for all LibOSes (it should expose bugs). For those LibOSes which the CI does not pass, disable them and file an issue to keep track of.

File: examples/tcp-close/client.rs [hunk 5]
```diff
@@ -150,4 +151,124 @@ impl TcpClient {
...
+                        qr_opcode => panic!("unexpected result (qr_opcode={:?})", qr_opcode),
+                    }
+                },
+                demi_opcode_t::DEMI_OPC_FAILED => panic!("operation failed (qr_ret={:?})", qr.qr_ret),
+                qr_opcode => panic!("unexpected result (qr_opcode={:?})", qr_opcode),
+            }
+
+            // Close TCP socket.
+            self.libos.close(sockqd)?;
+        }
```
Comment (iyzhang) [inline examples/tcp-close/client.rs:202]: Might be worth adding an error handling path that makes sure the TCP socket is closed even if we panic.

**Review by iyzhang — APPROVED** [review]
LGTM

**Review by ppenna — CHANGES_REQUESTED** [review]

File: tools/demikernel_ci.py [hunk 6]
```diff
@@ -353,13 +354,23 @@ def run_pipeline(
                                                                log_directory)
                 # TCP Close (optional)
                 if test_system == "all" or test_system == "tcp_close":
-                    if libos != "catnip" and libos != "catpowder":
+                    if libos == "catnap":
```
[the remark this answers is not in the document]
Comment (anandbonde) [reply tools/demikernel_ci.py]: Done

File: examples/tcp-close/server.rs [hunk 7]
```diff
@@ -175,6 +175,44 @@ impl TcpServer {
...
+            if let Some(nclients) = nclients {
+                if self.clients_closed >= nclients {
+                    // Sanity check that all connections have been closed.
+                    const ERR_MSG: &str = "there should be no clients connected, but there are";
+                    assert_eq!(self.clients.len(), 0, "{}", ERR_MSG);
+                    break;
+                }
+            }
+
+            let (_index, qr): (usize, demi_qresult_t) = self.libos.wait_any(&self.qts, None)?;
```
Comment (ppenna) [inline examples/tcp-close/server.rs:195]: We are missing to `remove()` the completed queue token here.

**Review by ppenna — APPROVED** [review]
LGTM once the issue on `remove()` is fixed.

File: examples/tcp-close/server.rs (hunk 7 above)
Comment (anandbonde) [reply examples/tcp-close/server.rs]: Done

**Pull request merged**